domain-models-runtime/src/main/java/org/folio/rest/persist/PgUtil.java

```diff
@@ -193,6 +225,147 @@ static String message(Throwable t) {
+      }
+      return respond422(respond422, field, uuid, INVALID_UUID);
+    } catch (Exception e) {
+      return response(e, failResponseMethod);
+    }
+  }
+
+  static Future<Response> responseForeignKeyViolation(String table, String id, PgExceptionFacade pgException,
+      Method response422method, Method valueMethod, Method failResponseMethod) {
+    try {
+      String detail = pgException.getDetail();
+      Matcher matcher = KEY_NOT_PRESENT_PATTERN.matcher(detail);
+      if (matcher.find()) {
+        String field = matcher.group(1);
+        String value = matcher.group(2);
+        String refTable = matcher.group(3);
+        String message = "Cannot set " + table + "." + field + " = " + value
+            + " because it does not exist in " + refTable + ".id.";
+        if (response422method == null) {
+          return response(message, valueMethod, failResponseMethod);
+        }
+        return respond422(response422method, table + "." + field, value, message);
+      }
+
+      matcher = KEY_STILL_REFERENCED_PATTERN.matcher(detail);
+      if (matcher.find()) {
+        String field = matcher.group(1);
+        String value = matcher.group(2);
+        String refTable = matcher.group(3);
+        String message = "Cannot delete " + table + "." + field + " = " + value
+            + " because id is still referenced from table " + refTable + ".";
+        if (response422method == null) {
+          return response(message, valueMethod, failResponseMethod);
+        }
+        return respond422(response422method, table + "." + field, value, message);
+      }
+
+      String message = pgException.getMessage() + " " + detail;
+      if (response422method == null) {
+        return response(message, valueMethod, failResponseMethod);
+      }
+      return respond422(response422method, table, id, message);
+    } catch (Exception e) {
+      return response(e, failResponseMethod);
+    }
+  }
+
+  static Future<Response> responseUniqueViolation(String table, String id, PgExceptionFacade pgException,
+      Method response422method, Method valueMethod, Method failResponseMethod) {
+    try {
+      String detail = pgException.getDetail();
+      Matcher matcher = KEY_ALREADY_EXISTS_PATTERN.matcher(detail);
+      if (! matcher.find()) {
+        detail = pgException.getMessage() + " " + detail;
+        if (response422method == null) {
+          return response(detail, valueMethod, failResponseMethod);
+        }
+        return respond422(response422method, table, id, detail);
+      }
+      String key = matcher.group(1);
+      String value = matcher.group(2);
+      String message = key + " value already exists in table " + table + ": " + value;
+      if (response422method == null) {
+        return response(message, valueMethod, failResponseMethod);
+      }
+      return respond422(response422method, key, value, message);
+    } catch (Exception e) {
+      return response(e, failResponseMethod);
+    }
+  }
+
```

Comment:
And two more identical looking functions here.. Why not just having a simple translation table of the Postgres IDs  and a map to the status?

Reply:
One Postgres ID can have several status messages, for example 23503 is the same for KEY_NOT_PRESENT and KEY_STILL_REFERENCED.

Reply:
Of course... And we'll have to maintain that mapping from Postgres to human readable messages or some of them for the future.. which is just one extra thing we'll have to update if Postgres changes behavior on this etc.. I really hope the users will enjoy this.. If we have a condition that a UI can react to .. A multilingual UI will have to map yet again.. Which is why an error code (NOT HTTP error code), is better for this.

Reply:
I think you should just merge this anyway.